ValveResourceFormat/KeyValues/KeyValues3.cs

```diff
@@ -0,0 +1,526 @@
+﻿/*
+ * KeyValues3 class.
+ * This class reads in Valve KV3 files and stores them in its datastructure.
+ *  
+ * Interface:
+ *  KVFile file = KV3Reader.ParseKVFile( fileName );
+ *  String fileString = file.Serialize();
+ *  
+ * TODO:
+ *  - Test some more and find the bugs.
+ *  - Revisit state machine if bugs require it.
+ *  - Improve KVFile interface.
+ *  
+ * Author: Perry - https://github.com/Perryvw
+ * 
+ */
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.IO;
+using System.Globalization;
+
+namespace ValveResourceFormat.KeyValues
+{
+    public static class KeyValues3
+    {
+        private enum State
+        {
+            HEADER,
+            SEEK_VALUE,
+            PROP_NAME,
+            VALUE_STRUCT,
+            VALUE_ARRAY,
+            VALUE_STRING,
+            VALUE_STRING_MULTI,
+            VALUE_NUMBER,
+            VALUE_FLAGGED,
+            COMMENT,
+            COMMENT_BLOCK
+        }
+
+        private class Parser
+        {
+            public StreamReader fileStream;
+
+            public KVObject root = null;
+
+            public String currentName;
+            public StringBuilder currentString;
+
+            public char previousChar;
+            public Queue<char> charBuffer;
+
+            public Stack<KVObject> objStack;
+            public Stack<State> stateStack;
+
+            public Parser()
+            {
+                //Initialise datastructures
+                objStack = new Stack<KVObject>();
+                stateStack = new Stack<State>();
+                stateStack.Push(State.HEADER);
+
+                root = new KVObject("root");
+                objStack.Push(root);
+
+                previousChar = '\0';
+                charBuffer = new Queue<char>();
+
+                currentString = new StringBuilder();
+            }
+        }
+
+        private const String numerals = "-0123456789.";
+
+        public static KVFile ParseKVFile(string filename)
+        {
+            using (var fileStream = new FileStream(filename, FileMode.Open, FileAccess.Read))
+            {
+                return ParseKVFile(fileStream);
+            }
+        }
+
+        public static KVFile ParseKVFile(Stream fileStream)
+        {
+            var parser = new Parser();
+
+            //Open stream reader
+            parser.fileStream = new StreamReader(fileStream);
```

Comment:
This seems like `fileStream` should be an immutable field set by the constructor. If it really really really should be a mutable property, make it PascalCase, not camelCase.